OSX elimination of IconRef, fully native Bitmap/Mask implementation

csomor opened the pull request:
Removing obsolete IconRef (classic mac) implementation, switching to bitmap implementation which does everything using CGBitmapContext. Right now for testing reasons there's still a switch to turn the 'old' implementation by setting `wxOSX_USE_ICONREF` to 1 and `wxOSX_BITMAP_NATIVE_ACCESS` to 0

csomor committed:
Fixing creation from CGContextRef

Our native buffer must be updated as well, otherwise RawAccess will fail

csomor committed:
Update deprecation in docs

csomor committed:
Initial Commit of full native bitmap macOS implementation

Right now, to better test regressions, with dual implementation

vadz commented on include/wx/osx/bitmap.h:
```diff
@@ -64,20 +62,36 @@ class WXDLLIMPEXP_CORE wxMask: public wxObject
     // a 8 bit depth mask
     void* GetRawAccess() const;
-    int GetBytesPerRow() const { return m_bytesPerRow ; }
+    int GetBytesPerRow() const;
```
I think those should have `OSX` prefix or be declared in `wxMaskBase` (and implemented in the existing `wxMask` implementations deriving from it) to make it more clear that they're not part of the public API or, alternatively, to make them part of the public API.

Otherwise it will be too easy to write code which compiles with wxMac, but not any other ports.

vadz commented on include/wx/osx/setup0.h:
```diff
@@ -1622,6 +1622,13 @@
 #define wxUSE_EPOLL_DISPATCHER 0
+
+// set to 1 if you have older code that still needs icon refs
+#define wxOSX_USE_ICONREF 0
```
If we intend to keep these options (i.e. if they're not going to be removed before/soon after this PR is merged), we also need to check for them in `include/wx/osx/chkconf.h`.

vadz commented on src/osx/carbon/utilscocoa.mm:
```diff
@@ -203,10 +203,23 @@ wxBitmap wxOSXCreateSystemBitmap(const wxString& name, const wxString &client, c
+    NSImage* nsimage = wxOSXGetSystemImage(name);
+    if ( nsimage )
+    {
+        // if ( sizeHint != wxDefaultSize )
```
It would be be better to avoid checking in commented out code.

csomor replied:
Yes, I'll do so, I'm not sure yet on which level a resize (if at all) should occur

vadz replied:
FWIW, I think that if it's possible not to do it all, it would be preferable.

vadz commented on src/osx/core/dcmemory.cpp:
```diff
@@ -65,14 +65,18 @@ void wxMemoryDCImpl::DoSelect( const wxBitmap& bitmap )
-        m_selected.BeginRawAccess() ;
+        wxASSERT_MSG( !bitmap.GetSelectedInto() ||
+                     (bitmap.GetSelectedInto() == GetOwner()),
+                     wxT("Bitmap is selected in another wxMemoryDC, delete the first wxMemoryDC or use SelectObject(NULL)") );
```
Please avoid using `wxT()` in the new code. In all but the most performance-critical places it's ok to omit it entirely, otherwise you can use `wxS()` to avoid an extra conversion, but `wxT()` is never useful.

csomor replied:
done

vadz commented on src/osx/core/bitmap.cpp:
```diff
@@ -62,19 +62,33 @@ class WXDLLEXPORT wxBitmapRefData: public wxGDIRefData
+    int GetBytesPerRow() const { return CGBitmapContextGetBytesPerRow(m_hBitmap); }
+    bool HasAlpha() const {
+        CGImageAlphaInfo alpha = CGBitmapContextGetAlphaInfo(m_hBitmap);
+        return !( alpha == kCGImageAlphaNone || alpha == kCGImageAlphaNoneSkipFirst || alpha == kCGImageAlphaNoneSkipLast) ;
```
It would be nice to avoid overlong (>> 80 characters) lines, they make side-by-side diffs less readable.

csomor replied:
Thanks, done

vadz commented on src/osx/core/bitmap.cpp:
```diff
@@ -62,19 +62,33 @@ class WXDLLEXPORT wxBitmapRefData: public wxGDIRefData
+    int GetHeight() const { return CGBitmapContextGetHeight(m_hBitmap); }
+    int GetDepth() const { return CGBitmapContextGetBitsPerPixel(m_hBitmap); }
+    int GetBytesPerRow() const { return CGBitmapContextGetBytesPerRow(m_hBitmap); }
+    bool HasAlpha() const {
```
Very minor, but in our style the `{` should be on its own line.

vadz commented on src/osx/core/bitmap.cpp:
```diff
@@ -364,57 +342,91 @@ bool wxBitmapRefData::Create(CGContextRef context)
+#endif
+    wxASSERT_MSG( hBitmap , wxT("Unable to create CGBitmapContext context") ) ;
+    CGContextTranslateCTM( hBitmap, 0,  GetHeight() );
+    CGContextScaleCTM( hBitmap, 1*GetScaleFactor(), -1*GetScaleFactor() );
```
Minor, but why `1*`? I find this more confusing than helpful...

csomor replied:
True, changed

vadz commented on src/osx/core/bitmap.cpp:
```diff
@@ -364,57 +342,91 @@ bool wxBitmapRefData::Create(CGContextRef context)
+{
+    wxCHECK_MSG( IsOk(), NULL , wxT("invalid bitmap") ) ;
+    wxASSERT_MSG( m_rawAccessCount == 1, "Direct write access must be within Begin/EndRawAccess" ) ;
+#if !wxOSX_BITMAP_NATIVE_ACCESS
```
Even if it's minimal, why duplicate the code already present above here? I'd write this as

    return const_cast<void*>(const_cast<const wxBitmapRefData*>(this)->GetRawAccess());

csomor replied:
Thanks, done

vadz reviewed (approved):
Thanks a lot for modernizing this!

All in all I don't have any real comments but it would obviously be even simpler if we could drop the new `wxOSX_USE_XXX` options and just always use the new code. Unless we really need to preserve the old behaviour in some special cases, I'd strongly prefer to do it before even merging this PR.

Thanks again!

csomor commented:
Hi Vadim, thanks for looking into things, yes, before merge the `wxOSX_BITMAP_NATIVE_ACCESS` will be removed, at least I planned so. I only left it in to allow for easier regression testing in case of bugs. But I think I'll still have to maintain the `wxOSX_USE_ICONREF`, because that's part of the public API, although not documented, or not ?

csomor commented:
`wxOSX_USE_ICONREF` is right now used in two ways, first as a switch to turn IconRef usage off completely, and then to use IconRef vs NSImage as implementation base (allowing for regression tests). So if no problems arise with the NSImage use, we could still support IconRef externally via the conversion.

vadz commented:
Sorry, I didn't even realize this. But I blame not just myself (as usual), but also the fact that the code is pretty confusing right now. I think we really need to check for `wxOSX_USE_ICONREF` in 2 places at most (instead of a dozen or more where it is done now): around `wxIconRefIconData` definition and when deciding which kind of `wxIconData` to create. Is there anything preventing us from doing it like this?

BTW, to help with code understanding, I think it would be useful to add `// wxOSX_USE_ICONREF` or `// !wxOSX_USE_ICONREF` comments after the corresponding `#endif`s, at least if they're more than a few lines removed from the matching `#if` or `#else`.

csomor commented:
Two additional Questions about removing/deprecating:
- I thought about removing SetWidth/Height/Depth for `wxIcon` as well ?
- I've get SetOk on `wxBitmap` and wxIcon and when using native data would not need them at all, no other ports seem to have them. How about direct removal ?

csomor commented:
OK, sorry for the confusion, I'll do a clean up and differentiate between the two situations in a cleaner way

vadz commented:
> I thought about removing SetWidth/Height/Depth for `wxIcon` as well ?

If we deprecate them for `wxBitmap`, it makes at least as much sense to do it for `wxIcon` too.

> I've get `SetOk` on wxBitmap and wxIcon and when using native data would not need them at all, no other ports seem to have them. How about direct removal ?

I think it should be fine, nobody seems to be using this one.

csomor committed:
Fixes and Cleanup

csomor committed:
only use native NSImage dimension

csomor commented:
The next question, related to [trac 18146](https://trac.wxwidgets.org/ticket/18146#comment:45) is the position of `wxIcon` vs `wxBitmap`. `wxIcon` is now based on `NSImage`, and since it does not have to provide direct access to the representation bitmap it can keep all appearance support advantages. So it would be the preferred class when using NSImage on OSX IMHO. 
But the current art provider implementation seems to favour bitmaps, even when `GetIcon` is called, it uses `GetBitmap`, which - as long as a bitmap is return from any provider - will be used before the icon bundles will even be considered ...

vadz commented:
Sorry, what do you mean by "position"? I.e. I'm not sure what is the question here?

And, whatever it is, it would be helpful if you could please explain in a few words (if it's possible) what is the difference between `wxBitmap` and `wxIcon` from macOS point of view? E.g. for MSW, an icon is just a bitmap with a mask (and, just to be complete, a cursor is an icon with a hotspot). But what about macOS, what's the real difference between icons and bitmaps there?

csomor commented:
OK, `wxBitmap` and `wxIcon` both represent graphics that can be drawn on a `wxDC`. `wxBitmap` offers methods for directly accessing the pixel map data, but `wxIcon` does not - unless converted to a `wxBitmap`. This makes `wxIcon` use an `NSImage` as native representation, which may very late, only at  draw-time decide what representation exactly will be drawn and how. While for a `wxBitmap` I must have / convert into a `CGBitmapContext` and therefore fixate the look in a pixel map, it cannot be adapted to the current appearance etc. later.

csomor commented:
ok, I should be able to support that, 
this would make wxIcon more or less a subset of wxBitmap on mac, didn't we have that before: ports where wxIcon is kind of a wxBitmap ?

vadz commented:
I'm late, so I can only say that I agree with Vaclav: ideal would be to change the internal representation only when necessary and use `NSImage` for as long as possible in `wxBitmap` too, as objects of this class are at least as often used as (small) icons as (possibly huge) images.

It's unfortunate that the same class is used for both, of course, but I think by now it's simpler to adapt the implementation than try to change the API...

csomor committed:
fixing graphics state stack error

csomor committed:
using newer API

csomor committed:
using NSImage as first-class native representation

If given an NSImage this will be used, unless raw bitmap data access is needed, lazily converting to a CGBitmapContext at that moment

csomor committed:
Direct NSImage drawing

csomor committed:
switching to generic icon for macOS

csomor committed:
new ICNS Handler

csomor committed:
forgotten commit of header change for generic icon

csomor committed:
use direct NSImage drawing commands

csomor committed:
Removing non-native bitmap code, using common code for iOS and macOS

csomor committed:
emoving non-native bitmap code, using common code for iOS and macOS

second part, sources were missing

csomor committed:
optimize handling of high-res native images

csomor committed:
makefile change using generic instead of osx icon file

csomor committed:
further cleanup

csomor committed:
Merge branch 'master' into osx_noIconRef

csomor committed:
collision fix

vadz commented:
Stefan, would you like me to merge this if it's ready? I'd like to do it after rebasing this on the latest master to get rid of the master merge.

vslavik commented:
> Hi @vslavik , could you please test whether the current state still works for your app, thanks

I did after your Thursday commits and re-tested now and everything seems fine. 

The only thing that doesn't work is correct rendering of template images in `wxStaticBitmap` — but that didn't work before either and [apparently](https://stackoverflow.com/questions/17148134/nsimage-drawinrect-and-templateimages) can only be fixed by writing a native implementation using `NSImageView`...

csomor commented:
> I did after your Thursday commits and re-tested now and everything seems fine.
> 
> The only thing that doesn't work is correct rendering of template images in `wxStaticBitmap` — but that didn't work before either and [apparently](https://stackoverflow.com/questions/17148134/nsimage-drawinrect-and-templateimages) can only be fixed by writing a native implementation using `NSImageView`...

@vslavik Perfect, thanks, would you have a sample at hand for the problem ?
then yes @vadz please go ahead (and tell me sometime what I did this time to make things difficult again)

vslavik commented:
> @vslavik Perfect, thanks, would you have a sample at hand for the problem ?

You mean for template images? They're simply rendered as if they weren't templates at all (i.e. typically black as opposed to gray dynamically chosen for the background). Here's a particularly nasty example, but the same difference exists in light mode:

![image](https://user-images.githubusercontent.com/145881/45929333-5f2eed80-bf50-11e8-9c5c-d9be1e7a121f.png)

This surprised me too, but template handling is something that is done at the controls level by `NSImageView` or `NSButton`.

vadz commented:
@csomor Nothing really difficult, but I just wanted to avoid unnecessary merge with master, I think it's preferable to rebase the branch on master instead in such cases.

In any case, I've merged it now, without any changes. I couldn't update `docs/changes.txt` as I don't really know what to put there, i.e. what are the user-visible effects of this change (if any?), so it would be great if you could please do. TIA!

csomor commented:
> You mean for template images? They're simply rendered as if they weren't templates at all (i.e. typically black as opposed to gray dynamically chosen for the background). Here's a particularly nasty example, but the same difference exists in light mode:
(...)
> This surprised me too, but template handling is something that is done at the controls level by `NSImageView` or `NSButton`.

I've tried using some etched drawing code, and this really looks nice for bigger images, but for small symbols it's not usable. I'll try a native implementation with NSImageView next.

vslavik commented:
> > Hi @vslavik , could you please test whether the current state still works for your app, thanks
>
> I did after your Thursday commits and re-tested now and everything seems fine.

I think I found a regression after all, in handling of stock icons (previously handled by `LoadIconFromSystemResource`, but I admit I'm a bit confused about which parts in git are used currently and which aren't). Reproducible with e.g. this, which tries to load `wxIcon(wxICON_WARNING, wxBITMAP_TYPE_ICON_RESOURCE)` and fails:
```cxx
wxArtProvider::GetBitmap(wxART_WARNING, wxART_MENU);
```

csomor commented:
I've checked with version 3.1.1 
```c++
wxArtProvider::GetBitmap(wxART_WARNING, wxART_MENU);
```
does not deliver this icon, due to the fact that first all bitmap providers are looked up, as I mentioned in a post about 3 weeks ago above. But using 
```c++
wxIcon("wxICON_WARNING", wxBITMAP_TYPE_ICON_RESOURCE)
```
gives the native one, and I'll add the code to do so on master, probably with a `LoadFile` extension on the `icns` Handler

csomor commented:
Hi @vslavik, did the commit improve things ?

vslavik commented:
@csomor I was *just* testing it — yes, it does (thanks!), although I think it's still subtly different from how it behaved before regarding icon sizes. See e.g. assert window:

<img width="650" alt="screen shot 2018-10-10 at 11 37 40" src="https://user-images.githubusercontent.com/145881/46727560-2053a400-cc81-11e8-8b32-281db45e545b.png">

I'm not actually sure it's an issue with icons loading and not the assert window, though.

csomor commented:
great, thanks, 
yes, size, that's true, right now I don't resize and get the maximum resolution, resizing would not be a problem, though, the question is when (and where) to do it - or not
